### PR TITLE
Perf improvements to the ipynb serializer

### LINKED ISSUE #129370: Avoid unnecessary string join and split
https://github.com/microsoft/vscode/blob/4e5d793a7828f1baf62e9e06ce5137023bd17c06/extensions/ipynb/src/helpers.ts#L364-L377

Currently the stream output conversion (from VS Code types to Jupyter) does unnecessary V8 string concatenation and split, which slows down the conversion (using more memory and gc):

* Stream output is either `CellOutputMimeTypes.stderr` or `CellOutputMimeTypes.stdout`, so `convertOutputMimeToJupyterOutput` will always return `string`. Using `prev.concat(curr)` will keep creating arrays
* `splitMultilineString(outputs.join(''))` can slow down the process significantly. It firstly joins all the string, and then split by line breaks, this will trigger v8 to flatten the concatenated string (`outputs.join('')`) and double the memory usage.

We can probably run `splitMultilineString` on each output and concatenate last line of each output with the first line of next output (split first, then join).

### PR DESCRIPTION
**Pull request by DonJayamanne**
This PR fixes #129370

### CODE REVIEW

File: extensions/ipynb/src/deserializers.ts [hunk 1]
```diff
@@ -61,8 +69,19 @@ function sortOutputItemsBasedOnDisplayOrder(outputItems: NotebookCellOutputItem[
 			}
 			return compareWith.startsWith(value);
 		};
-		const indexOfMimeTypeA = orderOfMimeTypes.findIndex(mime => isMimeTypeMatch(outputItemA.mime, mime));
-		const indexOfMimeTypeB = orderOfMimeTypes.findIndex(mime => isMimeTypeMatch(outputItemB.mime, mime));
+		let indexOfMimeTypeA = orderOfMimeTypes.findIndex((mime) => isMimeTypeMatch(mime, outputItemA.mime));
```
Comment (DonJayamanne) [inline extensions/ipynb/src/deserializers.ts:72]: Added something that was fixed in Jupyter extension (some how this was missing here).
I've added the changes/fixes here along with the corresponding tests.

File: extensions/ipynb/src/serializers.ts [hunk 2]
```diff
@@ -94,7 +94,7 @@ function translateCellDisplayOutput(output: NotebookCellOutput): JupyterOutput {
 		case 'display_data': {
 			result = {
 				output_type: 'display_data',
-				data: output.items.reduceRight((prev: any, curr) => {
```
Comment (DonJayamanne) [inline extensions/ipynb/src/serializers.ts:97]: Not sure why it was `reduceRight`, it should be `reduce` (fix everywhere else)

File: extensions/ipynb/src/serializers.ts [hunk 3]
```diff
@@ -224,17 +224,33 @@ type JupyterOutput =
...
 function convertStreamOutput(output: NotebookCellOutput): JupyterOutput {
-	const outputs = output.items
+	const outputs: string[] = [];
+	output.items
 		.filter((opit) => opit.mime === CellOutputMimeTypes.stderr || opit.mime === CellOutputMimeTypes.stdout)
 		.map((opit) => convertOutputMimeToJupyterOutput(opit.mime, opit.data as Uint8Array) as string)
-		.reduceRight<string[]>((prev, curr) => prev.concat(curr), []);
+		.forEach(value => {
+			// Ensure each line is a seprate entry in an array (ending with \n).
+			const lines = value.split('\n');
```
Comment (DonJayamanne) [inline extensions/ipynb/src/serializers.ts:233]: Made the changes as per the comments here https://github.com/microsoft/vscode/issues/129370#issue-952511326

File: extensions/ipynb/src/test/serializers.test.ts [hunk 4]
```diff
@@ -391,4 +466,131 @@ suite('ipynb serializer', () => {
 			});
 		});
 	});
+
+	suite('Output Order', () => {
+		test('Verify order of outputs', async () => {
+			const dataAndExpectedOrder: { output: nbformat.IDisplayData; expectedMimeTypesOrder: string[] }[] = [
```
Comment (DonJayamanne) [inline extensions/ipynb/src/test/serializers.test.ts:472]: Tests for a fix that was missing

**Review by rchiodo — APPROVED** [review]

File: extensions/ipynb/package.json [hunk 5]
```diff
@@ -64,8 +64,7 @@
     "watch": "npx gulp watch-extension:ipynb"
   },
 	"dependencies": {
-		"@enonic/fnv-plus": "^1.3.0",
-		"detect-indent": "^6.0.0"
```
Comment (DonJayamanne) [inline extensions/ipynb/package.json:68]: Removed this npm package & detecting indents ourselves

File: extensions/ipynb/src/notebookSerializer.ts [hunk 6]
```diff
@@ -94,3 +93,17 @@ export class NotebookSerializer implements vscode.NotebookSerializer {
 		return JSON.stringify(notebookContent, undefined, indentAmount);
 	}
 }
+
+export function detectIndent(jsonString: string) {
```
Comment (DonJayamanne) [inline extensions/ipynb/src/notebookSerializer.ts:97]: Forgot to complete this in the first submission of the PR.

File: extensions/ipynb/package.json (hunk 5 above)
Comment (rebornix) [reply extensions/ipynb/package.json:68]: 🎉

File: extensions/ipynb/src/deserializers.ts [hunk 7]
```diff
@@ -53,6 +53,14 @@ const orderOfMimeTypes = [
 	'text/plain'
 ];
 
+function isEmptyVendoredMimeType(outputItem: NotebookCellOutputItem) {
+	if (outputItem.mime.startsWith('application/vnd.')) {
+		try {
+			return Buffer.from(outputItem.data).toString().length === 0;
```
Comment (rebornix) [inline extensions/ipynb/src/deserializers.ts:59]: Is this the same as checking `outputItem.data.byteLength === 0` or maybe to be safe we can check `outputItem.data.byteLength === 0 || Buffer.from(outputItem.data).toString().length === 0`

File: extensions/ipynb/src/notebookSerializer.ts (hunk 6 above)
Comment (rebornix) [reply extensions/ipynb/src/notebookSerializer.ts:97]: Though I like that we get rid of `detect-indent` a lot but I'm a bit worried if the algorithm here works the same as `detect-indent` for common cases. Not sure how much coverage we can have with tests 

Another algorithm to compare against is the one we have for monaco: https://github.com/microsoft/vscode/blob/79b66ea6ff73559c7481eb3ecde1cd3f5b4dd680/src/vs/editor/common/model/indentationGuesser.ts#L105-L1

File: extensions/ipynb/src/serializers.ts [hunk 8]
```diff
@@ -224,17 +224,33 @@ type JupyterOutput =
...
+			// If the last item in `outputs` is not empty and the first item in `lines` is not empty, then concate them.
+			// As they are part of the same line.
+			if (outputs.length && lines.length && lines[0].length > 0) {
+				outputs[outputs.length - 1] = `${outputs[outputs.length - 1]}${lines.shift()!}`;
+			}
+			while (lines.length) {
+				outputs.push(lines.shift()!);
```
Comment (rebornix) [inline extensions/ipynb/src/serializers.ts:240]: `shift` will reindex the array, we can do a simple `outputs.push(lines[i])`

File: extensions/ipynb/src/serializers.ts (hunk 3 above)
Comment (rebornix) [reply extensions/ipynb/src/serializers.ts:233]: for the split, we do want to handle CRLF?

File: extensions/ipynb/src/deserializers.ts [hunk 9]
```diff
@@ -53,6 +53,14 @@ const orderOfMimeTypes = [
...
+		} catch { }
+	}
+	return false;
+}
 function sortOutputItemsBasedOnDisplayOrder(outputItems: NotebookCellOutputItem[]): NotebookCellOutputItem[] {
 	return outputItems.sort((outputItemA, outputItemB) => {
```
Comment (rebornix) [inline extensions/ipynb/src/deserializers.ts:65]: Considering that checking the index of an output item is not trivial, doing a sort directly here will lead to running `orderOfMimeTypes.findIndex` multiple times.

Maybe we can do `outputItems.map(item => ({ item, index: findIndexOfMimeType(item) }).sort((a, b) => a -b)`

**Review by rebornix — CHANGES_REQUESTED** [review]
:clap: for the multi line stream improvement, only a few suggestions of how to improve the sorting.

Regarding indentation detection, I like that we can drop the external dependency but not sure how confident we are with the behavior change.

File: extensions/ipynb/src/serializers.ts (hunk 3 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/serializers.ts:233]: I don't think we need to (this is how we used to do it & based on what I know this is what Jupyter also does, - code was borrowed from Jupyter)

Comment (DonJayamanne) [reply extensions/ipynb/src/serializers.ts:233]: @rchiodo /cc

File: extensions/ipynb/src/notebookSerializer.ts (hunk 6 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/notebookSerializer.ts:97]: Not exactly sure what you;re asking here. Even if we had a mix of tabs & spaces this should work as `trimStart` would strip both of them.
And looking at the source you're provided, its ensuring spaces & tabs are both preserved.
E.g. if the json indentation is `\t    `, then we'll identify the tab+spaces & preserve that.

Comment (DonJayamanne) [reply extensions/ipynb/src/notebookSerializer.ts:97]: I could optionally bring back the npm package & only provide the first 10-20 characters of the string. 
* That will address the issue with any edge cases
* Addresses perf issues (not providing the entire string, only part of it).

Thoughts?
@rebornix @roblourens /cc

Comment (rebornix) [reply extensions/ipynb/src/notebookSerializer.ts:97]: The problem IMHO is unformated document which contains both 2 spaces tab and 4 spaces tabs. 
The catch is indentation detection is always a `guess` and I think both `detect-indent` and monaco's one will detect the indentation of every line and find the most common indentation in the document. For example, if a document's first line uses 2 spaces but the rest of the document use 4 spaces, then the indentation from the algorithm would be 4, other than 2, while checking the first 20 characters will very likely give us 2.

In the core, we only detect indentation of the first 10,000 lines, maybe we can do the same for `detect-indent` (tested locally and it does help with the perf for large files).

Comment (DonJayamanne) [reply extensions/ipynb/src/notebookSerializer.ts:97]: > contains both 2 spaces tab and 4 spaces tabs.
The catch is indentation detection is always a guess and I think both detect-indent

FYI - here we're not detecting indents entered by the user. 
We're detecting indents in the serialized JSON, not any user entered code. Hence your concerns are not really valid here. We're talking about the output of JSON.stringify (or equivalent in Python) - which is done by jupyter applicatino, not user input values

File: extensions/ipynb/src/notebookSerializer.ts [hunk 10]
```diff
@@ -94,3 +93,17 @@ export class NotebookSerializer implements vscode.NotebookSerializer {
...
+	// Lets just take the distance between the first `{` and the next non-white space character`, ignoring \r & \n
+	if (!jsonString.startsWith('{')) {
+		return '';
+	}
+	// We're only interested in a small part of the string.
+	// The assumption is that we won't have an indentation of 10, just around 5 or so.
+	jsonString = jsonString.substring(1, 10).replace(/\r?\n/g, '');
+	// first index of non white space is the indentation.
+	const firstPositionOfNonWhiteSpace = jsonString.length - jsonString.trimStart().length;
+	return jsonString.substring(0, firstPositionOfNonWhiteSpace);
```
Comment (roblourens) [inline extensions/ipynb/src/notebookSerializer.ts:108]: If the file is `{ "a": 1 }` I think that would be parsed as one space of indentation by this code, but there is no indentation

File: extensions/ipynb/src/test/serializers.test.ts [hunk 11]
```diff
@@ -107,6 +133,81 @@ suite('ipynb serializer', () => {
...
 			);
 		});
+		test('Multi-line Stream output', () => {
+			validateCellOutputTranslation(
+				[
+					{
+						name: 'stdout',
+						output_type: 'stream',
+						text: [
+							'Epoch 1/5\n',
```
Comment (roblourens) [inline extensions/ipynb/src/test/serializers.test.ts:143]: I think the tests would be more readable with simpler examples, that's just my personal preference though

**Review by roblourens — APPROVED** [review]

File: extensions/ipynb/src/test/serializers.test.ts (hunk 11 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/test/serializers.test.ts:143]: I thought of adding a more real world example, but will change to a simpler output.

File: extensions/ipynb/src/deserializers.ts (hunk 9 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/deserializers.ts:65]: Yup missed that

File: extensions/ipynb/src/test/serializers.test.ts (hunk 11 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/test/serializers.test.ts:143]: Done

**Review by rebornix — APPROVED** [review]
It looks great! 🚢

**Pull request merged**